<a href="https://colab.research.google.com/github/MuhammadMaazKhan/Machine-Learning/blob/main/4.Association%20Rule%20Learning/0.Apriori/apriori.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Apriori

## Importing the libraries

In [ ]:
!pip install apyori

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Data Preprocessing

Apriori expect particular format for dataset and unfortunately that format is not a pandas dataframe. it requires one by one full transactions list or movies watched list by one person.

so now we make transaction array , and in that array we filled the data one by one  full row (full transaction by one person)

**[str(dataset.values[i,j]) for j in range(0, 20)]**

 where i  = row , j = cols

 so we have 19 cols max in dataset thats why we take range (0,20)

 a certain **expectation of the future**, a primary function which we'll use to train our primary model, it is the fact that all the elements in the **list**
must be **strings.**



In [ ]:
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header = None)
transactions = []
for i in range(0, 7501):
  transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])

## Training the Apriori model on the dataset


**transactions**=transactions → Passes the list of transactions to Apriori.

**min_support**=0.003 → Minimum support threshold (frequency of an itemset appearing in transactions).

This means an item must be present in at least 0.3% of all transactions to be considered.

**min_confidence**=0.2 → Minimum confidence threshold (how often a rule is correct).

A rule is kept only if at least 20% of the time, when item X is bought, item Y is also bought.

**min_lift**=3 → Minimum lift threshold (strength of the association).

The rule must be at least 3 times more likely than random chance.

**min_length**=2, **max_length=2** → Limits the size of itemsets to exactly two items.

Ensures only pairwise product associations are found (e.g., “People who bought bread also bought butter”).

In [ ]:
from apyori import apriori
rules = apriori(transactions = transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2, max_length = 2)

## Visualising the results

### Displaying the first results coming directly from the output of the apriori function

In [ ]:
results = list(rules)

In [ ]:
results

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'honey', 'fromage blanc'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)])

item_base is based item mean "who bought light cream they also bought chicken"

in further steps we say **lhs** to item-base and **rhs** to items-add


Step 1: Extract Left-Hand Side (LHS)

**lhs = [tuple(result[2][0][0])[0] for result in results]**

result[2] → Accesses ordered_statistics.

[0][0] → Gets the items_base (left-hand side of the rule).

tuple(...) → Converts it into a tuple

✅ Example Output: ('light cream',)

tuple(...)[0] → extracts the first item which is name of lhs

✅ Example Output: "light cream"

Step 2: Extract Right-Hand Side (RHS)

**rhs = [tuple(result[2][0][1])[0] for result in results]**

Similar to LHS extraction, but gets items_add (right-hand side of the rule).

✅ Example Output: "chicken"

Step 3: Extract Support

**supports = [result[1] for result in results]**

result[1] → Directly gets the support value.

✅ Example Output: 0.0045

Step 4: Extract Confidence

**confidences = [result[2][0][2] for result in results]**

result[2][0][2] → Gets confidence value from ordered statistics.

✅ Example Output: 0.2906

Step 5: Extract Lift

**lifts = [result[2][0][3] for result in results]**

result[2][0][3] → Extracts lift value.

✅ Example Output: 4.8439

Step 6: Organizing into a DataFrame

**return list(zip(lhs, rhs, supports, confidences, lifts))**

**resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])**

zip(lhs, rhs, supports, confidences, lifts) → Combines all extracted data into a structured list.

pd.DataFrame(..., columns=['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift']) → Formats the list into a readable table.

✅ Example Output (DataFrame):

 |Left Hand Side | Right Hand Side | Support | Confidence | Lift|
 |---------------|-----------------|---------|------------|-----|
 |light cream    | chicken        | 0.0045   | 0.2906     | 4.84|  

### Putting the results well organised into a Pandas DataFrame

In [ ]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])

### Displaying the results non sorted

In [ ]:
resultsinDataFrame

Left Hand Side Right Hand Side   Support  Confidence      Lift
0           light cream         chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
2                 pasta        escalope  0.005866    0.372881  4.700812
3         fromage blanc           honey  0.003333    0.245098  5.164271
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
6           light cream       olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
8                 pasta          shrimp  0.005066    0.322034  4.506672

### Displaying the results sorted by descending lifts

In [ ]:
resultsinDataFrame.nlargest(n = 10, columns = 'Lift')

Left Hand Side Right Hand Side   Support  Confidence      Lift
3         fromage blanc           honey  0.003333    0.245098  5.164271
0           light cream         chicken  0.004533    0.290598  4.843951
2                 pasta        escalope  0.005866    0.372881  4.700812
8                 pasta          shrimp  0.005066    0.322034  4.506672
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
6           light cream       olive oil  0.003200    0.205128  3.114710

**Understanding the Columns (output)**

Left Hand Side (LHS) → The base product in the association (the trigger).

Right Hand Side (RHS) → The associated product (often bought together with LHS).

Support → The **percentage of transactions** that contain both products.

Example: 0.003333 means 0.33% of all transactions included fromage blanc and honey together.

Confidence → The **probability** that when LHS is bought, RHS is bought too.

Example: 0.245098 means 24.5% of customers who bought fromage blanc also bought honey.

Lift → The **strength of the association** (higher is better).

Example: 5.164271 means honey is 5.16 times more likely to be bought when fromage blanc is purchased, compared to random chance.